In [2]:
import pandas as pd
import numpy as np

In [37]:
Kindly note that the Data from january to june has all merged in the file named "travels_data.csv" and saved on the system

df = pd.read_csv(r'C:\Users\USER\Desktop\OCTAVE\travels_data.csv')

print('The number of rows and columns  in the dataframe are {} and {}'.format(df.shape[0], df.shape[1]))
df.head()

The number of rows and columns  in the dataframe are 3634338 and 109


Unnamed: 0  YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK     FL_DATE  \
0           0  2019        1      1             3            4  03/01/2019   
1           1  2019        1      1             4            5  04/01/2019   
2           2  2019        1      1             5            6  05/01/2019   
3           3  2019        1      1             6            7  06/01/2019   
4           4  2019        1      1             7            1  07/01/2019   

  OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID OP_CARRIER  ... DIV4_WHEELS_OFF  \
0                9E                  20363         9E  ...             0.0   
1                9E                  20363         9E  ...             0.0   
2                9E                  20363         9E  ...             0.0   
3                9E                  20363         9E  ...             0.0   
4                9E                  20363         9E  ...             0.0   

   DIV4_TAIL_NUM  DIV5_AIRPORT  DIV5_AIRPORT_ID  DIV5_AIRPORT_SEQ_ID  \
0            0.0           0.0              0.0                  0.0   
1            0.0           0.0              0.0                  0.0   
2            0.0           0.0              0.0                  0.0   
3            0.0           0.0              0.0                  0.0   
4            0.0           0.0              0.0                  0.0   

  DIV5_WHEELS_ON DIV5_TOTAL_GTIME DIV5_LONGEST_GTIME  DIV5_WHEELS_OFF  \
0            0.0              0.0                0.0              0.0   
1            0.0              0.0                0.0              0.0   
2            0.0              0.0                0.0              0.0   
3            0.0              0.0                0.0              0.0   
4            0.0              0.0                0.0              0.0   

  DIV5_TAIL_NUM  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  

[5 rows x 109 columns]

In [4]:
# filling the missing values with zero

df = df.fillna(0)
df.isna().sum()

Unnamed: 0            0
YEAR                  0
QUARTER               0
MONTH                 0
DAY_OF_MONTH          0
                     ..
DIV5_WHEELS_ON        0
DIV5_TOTAL_GTIME      0
DIV5_LONGEST_GTIME    0
DIV5_WHEELS_OFF       0
DIV5_TAIL_NUM         0
Length: 109, dtype: int64

In [40]:
print('The number of obervations for flights that are not cancelled is {}'.format(uncancelled_flight.shape[0]))

print('In subsequent Aanalysis the flights that are not cancelled will be denoted by "successful_flights" or "uncancelled_flight"')


The number of obervations for flights that are not cancelled is 3549066
In subsequent Aanalysis the flights that are not cancelled will be denoted by "successful_flights" or "uncancelled_flight"


### 1. What airline had the most number of flights in the first quarter of 2019 ?

In [41]:
uncancelled_flight =df.loc[df['CANCELLED']==0]

first_quater = uncancelled_flight.loc[uncancelled_flight['QUARTER']==1]


In [6]:
first_quater = first_quater[['OP_UNIQUE_CARRIER','FLIGHTS']]

first_quater = first_quater.groupby('OP_UNIQUE_CARRIER').sum()

airline_with_most_number_of_flights = first_quater.loc[first_quater['FLIGHTS']==first_quater['FLIGHTS'].max()]

airline_with_most_number_of_flights

FLIGHTS
OP_UNIQUE_CARRIER         
WN                  316404

### 2. What is the total number of Airline Travels in the first half of 2019 ? 

In [7]:

uncancelled_flight =df.loc[df['CANCELLED']==0]

number_of_Airline_Travels = uncancelled_flight['FLIGHTS'].sum()

print('The total number of Airline Travels in the first half of 2019 is {}'.format(number_of_Airline_Travels))

The total number of Airline Travels in the first half of 2019 is 3549066


### 3. What is the Total AirTime by all Airlines in May 2019 ?

In [8]:
may = df.loc[df['MONTH']==5]

Total_AirTime =may['AIR_TIME'].sum()

print('What is the Total AirTime by all Airlines in May 2019 is {}'.format(Total_AirTime))

What is the Total AirTime by all Airlines in May 2019 is 69297104.0


### 4. which is the most busiest city in April 2019

In [10]:
#first we check by the city with the largest number of flights entry

successful_flights = df.loc[df['CANCELLED']==0]

april = successful_flights.loc[successful_flights['MONTH']==4]

april = april[['DEST_CITY_NAME','FLIGHTS']]

april = april.groupby('DEST_CITY_NAME').sum()

max_city_travelled = april.loc[april['FLIGHTS'] == april['FLIGHTS'].max()]

max_city_travelled

FLIGHTS
DEST_CITY_NAME         
Chicago, IL       32978

In [11]:
#secondly we check by the city with the largest number of flights departure

april2 = successful_flights.loc[successful_flights['MONTH']==4]

april2 = april2[['ORIGIN_CITY_NAME','FLIGHTS']]

april2 = april2.groupby('ORIGIN_CITY_NAME').sum()

max_city_departed = april2.loc[april2['FLIGHTS'] == april2['FLIGHTS'].max()]

max_city_departed


FLIGHTS
ORIGIN_CITY_NAME         
Chicago, IL         32979

#### Since chicago has the highest number of flight arrivals from other cities and the highest number of Flight departure,

#### we can say that is the most busiest city

### 6.What airline covered the most distance in June 2019 ?

In [13]:

june_record = successful_flights.loc[successful_flights['MONTH']==6]

distance_cal = june_record.groupby('OP_UNIQUE_CARRIER').sum()

In [14]:
# we observe the airport distance and the diverted airport distance in cases of diverts

distance_cal = distance_cal[['DISTANCE','DIV_DISTANCE']]

# we add the two distances together since entry with no divert and diverts 
# that arrived scheduled airport before its divert are null

distance_cal['TOTAL_DISTANCE'] = distance_cal['DISTANCE'] + distance_cal['DIV_DISTANCE']

distance_cal.head()

DISTANCE  DIV_DISTANCE  TOTAL_DISTANCE
OP_UNIQUE_CARRIER                                        
9E                  9748411         452.0       9748863.0
AA                 77478118        9261.0      77487379.0
AS                 30631435        4474.0      30635909.0
B6                 26877337         844.0      26878181.0
DL                 79526511         513.0      79527024.0

In [15]:
airline_the_most_distance = distance_cal.loc[distance_cal['TOTAL_DISTANCE']==distance_cal['TOTAL_DISTANCE'].max()]
airline_the_most_distance

DISTANCE  DIV_DISTANCE  TOTAL_DISTANCE
OP_UNIQUE_CARRIER                                        
WN                 83654773       24618.0      83679391.0

### 7.What is the Tail Number of the aircraft that covered the most distance in Q2 of 2019 ?

In [16]:
# slice out the records for the second quarter
second_qtr= successful_flights.loc[successful_flights['QUARTER']==2]


# feature the distances and tail_number before grouping

distance = second_qtr[['TAIL_NUM','DISTANCE','DIV_DISTANCE']]

distance = distance.groupby('TAIL_NUM').sum() 

distance.head()

DISTANCE  DIV_DISTANCE
TAIL_NUM                        
215NV       283301           0.0
216NV       300892           0.0
217NV       262953           0.0
218NV       302323           0.0
219NV       272308           0.0

In [17]:
#  sum the distances

distance['TOTAL_DISTANCE']=distance['DISTANCE'] + distance['DIV_DISTANCE']
distance.head()

DISTANCE  DIV_DISTANCE  TOTAL_DISTANCE
TAIL_NUM                                        
215NV       283301           0.0        283301.0
216NV       300892           0.0        300892.0
217NV       262953           0.0        262953.0
218NV       302323           0.0        302323.0
219NV       272308           0.0        272308.0

In [18]:
Tail_num_max_distance_travelled = distance.loc[distance['TOTAL_DISTANCE']==distance['TOTAL_DISTANCE'].max()]
Tail_num_max_distance_travelled

DISTANCE  DIV_DISTANCE  TOTAL_DISTANCE
TAIL_NUM                                        
N934JB      598286           0.0        598286.0

### 8. Which state recorded the most delays in the 2nd Quarter of 2019 

In [26]:

second_quarter= successful_flights.loc[successful_flights['QUARTER']==2]

# slice the dataframe

delayed_records = second_qtr[['ORIGIN_STATE_NM','DEST_STATE_NM','DEP_DELAY_NEW','ARR_DELAY_NEW','DIV_ARR_DELAY']]
delayed_records.head()

ORIGIN_STATE_NM DEST_STATE_NM  DEP_DELAY_NEW  ARR_DELAY_NEW  \
1749234        New York       Florida           11.0            4.0   
1749235        New York       Florida            0.0            0.0   
1749236        New York       Florida            0.0            0.0   
1749237        New York       Florida           10.0           47.0   
1749238        New York       Florida           66.0           85.0   

         DIV_ARR_DELAY  
1749234            0.0  
1749235            0.0  
1749236            0.0  
1749237            0.0  
1749238            0.0

#### Checking for DESTINATION STATES

In [23]:
delay_records = delay_records.groupby('DEST_STATE_NM').sum()
delay_records.head()

DEP_DELAY_NEW  ARR_DELAY_NEW  DIV_ARR_DELAY  TOTAL_DELAYS
DEST_STATE_NM                                                           
Alabama             156206.0       163706.0         4657.0      324569.0
Alaska               74616.0        75349.0         3500.0      153465.0
Arizona             703505.0       683904.0        12432.0     1399841.0
Arkansas            115943.0       128499.0         5804.0      250246.0
California         2769521.0      2691577.0        39618.0     5500716.0

In [21]:
#sum the delays

delay_records['TOTAL_DELAYS']= delay_records['DEP_DELAY_NEW'] + delay_records['ARR_DELAY_NEW'] + delay_records['DIV_ARR_DELAY']


max_state_delays = delay_records.loc[delay_records['TOTAL_DELAYS']==delay_records['TOTAL_DELAYS'].max()]
max_state_delays

DEP_DELAY_NEW  ARR_DELAY_NEW  DIV_ARR_DELAY  TOTAL_DELAYS
DEST_STATE_NM                                                           
Texas              3502519.0      3731380.0       355436.0     7589335.0

### Checking for origin states

In [27]:
delay_records2 = delayed_records.groupby('ORIGIN_STATE_NM').sum()


delay_records2['TOTAL_DELAYS']= delay_records2['DEP_DELAY_NEW'] + delay_records2['ARR_DELAY_NEW'] + delay_records2['DIV_ARR_DELAY']


max_state_delays2 = delay_records2.loc[delay_records2['TOTAL_DELAYS']==delay_records2['TOTAL_DELAYS'].max()]
max_state_delays2


DEP_DELAY_NEW  ARR_DELAY_NEW  DIV_ARR_DELAY  TOTAL_DELAYS
ORIGIN_STATE_NM                                                           
Texas                3770427.0      3833069.0       152677.0     7756173.0

### 9. What is the daily average Taxi-out time in New York in February 2019 ?

In [28]:
february = df.loc[df['MONTH']==2]

New_york = february.loc[february['ORIGIN_STATE_NM']=="New York"]

New_york.shape[0]

29152

In [29]:
total_daily_observations = New_york.shape[0]

total_daily_records = New_york['TAXI_OUT'].sum()

Daily_average = total_daily_records/total_daily_observations

print('The daily average Taxi-out time in New York in February 2019 is {}'.format(Daily_average))

The daily average Taxi-out time in New York in February 2019 is 23.627675631174533


### 10. How many airlines with New York destinations were diverted in Q1 2019 ?

In [30]:
first_quarter= successful_flights.loc[successful_flights['QUARTER']==1]

new_york_dest = first_quarter.loc[first_quarter['DEST_STATE_NM']=="New York"]

#Slicing the dataframe to get records of diverts
new_york_diverts = new_york_dest.loc[new_york_dest['DIV_AIRPORT_LANDINGS']>0]


In [31]:
airlines = new_york_diverts['OP_UNIQUE_CARRIER'].unique()

total_airlines = len(airlines)

print('The total count of airlines with New York destinations which were diverted in Q1 2019 is {}'.format(total_airlines))

The total count of airlines with New York destinations which were diverted in Q1 2019 is 17


### 11. Which of these delays is more rampant in Washington in the first half of 2019

In [ ]:
# to get this we will check the number of times each delay type occured 

In [32]:
# slicing out the record for the delay causes

washington = df.loc[df['ORIGIN_STATE_NM']=="Washington"]

washington = washington[['ORIGIN_STATE_NM','CARRIER_DELAY','WEATHER_DELAY','NAS_DELAY','SECURITY_DELAY','LATE_AIRCRAFT_DELAY']]
washington.head()

ORIGIN_STATE_NM  CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
2327      Washington            0.0            0.0        0.0             0.0   
2328      Washington            0.0            0.0        0.0             0.0   
2329      Washington            0.0            0.0        0.0             0.0   
2330      Washington            0.0            0.0        0.0             0.0   
2331      Washington            0.0            0.0        0.0             0.0   

      LATE_AIRCRAFT_DELAY  
2327                  0.0  
2328                  0.0  
2329                  0.0  
2330                  0.0  
2331                  0.0

In [33]:
# checking for occurrences of CARRIER_DELAY

washington_CARRIER_DELAY = washington.loc[washington['CARRIER_DELAY']>0]

NO_CARRIER_DELAY_OCCURRENCES = washington_CARRIER_DELAY.shape[0]

print(f'NO_CARRIER_DELAY_OCCURRENCES = {NO_CARRIER_DELAY_OCCURRENCES}')

NO_CARRIER_DELAY_OCCURRENCES = 5265


In [34]:
# checking for occurrences of WEATHER_DELAY

washington_WEATHER_DELAY = washington.loc[washington['WEATHER_DELAY']>0]

NO_WEATHER_DELAY_OCCURRENCES = washington_WEATHER_DELAY.shape[0]

print(f'NO_WEATHER_DELAY_OCCURRENCES = {NO_WEATHER_DELAY_OCCURRENCES}')

NO_WEATHER_DELAY_OCCURRENCES = 827


In [35]:
# checking for occurrences of WEATHER_DELAY

washington_NAS_DELAY = washington.loc[washington['NAS_DELAY']>0]

NO_NAS_DELAY_OCCURRENCES = washington_NAS_DELAY.shape[0]

print(f'NO_NAS_DELAY_OCCURRENCES = {NO_NAS_DELAY_OCCURRENCES}')

NO_NAS_DELAY_OCCURRENCES = 8253


In [115]:
# checking for occurrences of SECURITY_DELAY


washington_SECURITY_DELAY = washington.loc[washington['SECURITY_DELAY']>0]

NO_SECURITY_DELAY_OCCURRENCES = washington_SECURITY_DELAY.shape[0]

print(f'NO_SECURITY_DELAY_OCCURRENCES = {NO_SECURITY_DELAY_OCCURRENCES}')

NO_SECURITY_DELAY_OCCURRENCES = 43


In [36]:
# checking for occurrences of LATE_AIRCRAFT_DELAY


washington_LATE_AIRCRAFT_DELAY = washington.loc[washington['LATE_AIRCRAFT_DELAY']>0]

NO_LATE_AIRCRAFT_DELAY_OCCURRENCES = washington_LATE_AIRCRAFT_DELAY.shape[0]

print(f'NO_LATE_AIRCRAFT_DELAY_OCCURRENCES = {NO_LATE_AIRCRAFT_DELAY_OCCURRENCES}')

NO_LATE_AIRCRAFT_DELAY_OCCURRENCES = 5161


#### The most rampant delay in Washington in the first half of 2019 is National Air System Delay with about 8253 occurrences

In [42]:
washington = df.loc[df['DEST_STATE_NM']=="Washington"]

washington = washington[['DEST_STATE_NM','CARRIER_DELAY','WEATHER_DELAY','NAS_DELAY','SECURITY_DELAY','LATE_AIRCRAFT_DELAY']]
washington.head()

DEST_STATE_NM  CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
2160    Washington            0.0            0.0        0.0             0.0   
2161    Washington            0.0            0.0        0.0             0.0   
2162    Washington            0.0            0.0        0.0             0.0   
2163    Washington            0.0            0.0        0.0             0.0   
2615    Washington            0.0            0.0        0.0             0.0   

      LATE_AIRCRAFT_DELAY  
2160                  0.0  
2161                  0.0  
2162                  0.0  
2163                  0.0  
2615                  0.0

Note: the same data above was calculated using the DESTINATION STATE NAME and the most occurrences still remain the 
National Air System Delay  
